### ISSUE TITLE
Media Library - Keyboard Navigation

### ISSUE THREAD

**Issue by jauyong** [GoogleForCreators/web-stories-wp#1305]
## Feature Description

As an author I want to be able to navigate the media library via keyboard

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria

1. Implement keyboard navigation. 
2. Keyboard should trigger infinite scroll loading for next page.

**Comment by diegovar** [comment]
By keyboard navigation, do you mean being able to use the Tab key to cycle between focused elements?

If not, can you elaborate on how the navigation would work?

**Comment by swissspidy** [comment]
Tab key wouldn't work. With the infinite scrolling in the media library, you would never be able to get out of it. So should be arrow keys. See discussion at #248 about this.

**Comment by samitron7** [comment]
@pbakaus Can you take the logic piece here since you're doing a holistic pass with kb navigation. I'll work to supply the focus state with Victor. Will update once we have the designs locked this Friday.

**Comment by pbakaus** [comment]
@samitron7 actually passing this over to Fontaine makes me most sense here - I don't think I know her Github handle, but I'll get an email thread going. My [A11Y doc](https://docs.google.com/document/d/15FHCukTDrWf7xYpPkBG3bGRgRQZeqzkCcglaNXgaw9s/edit#) focuses on general guidelines. The feature specific work still needs to happen in the feature specs.

**Comment by diegovar** [comment]
I think infinite scrolling and a11y are very much at odds, and from my research on the subject it's extremely hard if not impossible to build an infinite scrolling feed that eg: screen reader users can find usable. There's a few problems:
1) Users can't use tab to move around because as Pascal mentioned they would never be able to get out of the list (to navigate to other parts of the editor).
2) AFAIK you can't also suddenly ask them to use the Up/down/left/right keys when they happen to land in the list because they have no way of knowing that their method of input is supposed to change. So tabs should work.

I think that a good way around this is instead of automatically loading the next page, we should have a "Load more images" button at the bottom, which when clicked loads the next page (similar to what Google Search does). Then you keep on tabbing until you reach the end of the second page, click that button again, etc. WDYT @samitron7 ?

**Comment by swissspidy** [comment]
> AFAIK you can't also suddenly ask them to use the Up/down/left/right keys when they happen to land in the list because they have no way of knowing that their method of input is supposed to change. So tabs should work.

I disagree. You **can** inform users that they can interact with a list using arrow keys, and you _should_ do that. In fact, it's pretty common and encouraged. See ["Managing Focus"](https://www.w3.org/TR/wai-aria-1.1/#managingfocus) section of the ARIA spec:

> When the container or its active descendant has focus, the user may navigate through the container by pressing additional keys, such as the arrow keys, to change the currently active descendant. Any additional press of the main navigation key (generally the TAB key) will move out of the container to the next widget.

You cannot expect user to tab through 30 items before they can get to the next important area on the screen. The [grid widget](https://www.w3.org/TR/wai-aria-practices-1.1/#grid) might be a suitable way for us to announce that navigation style, but there are others too. [The examples](https://www.w3.org/TR/wai-aria-practices-1.1/examples/grid/LayoutGrids.html) list "a set of search results" as a possible use case.

You are right about the infinite scrolling part. It's difficult to get this right in an accessible way. WordPress core is even considering [removing infinite scrolling](https://core.trac.wordpress.org/ticket/50105) for its media library. Some of the issues mentioned there:

* it's impossible or very hard for keyboard users to reach content placed after an infinite scrolling region
* no audible feedback or instructions about how infinite scrolling works, the current and total number of items, or when new items get loaded

So my tl;dr is:

* Arrow keys for navigation
* Don't mess with tabs
* "load more" button instead of infinite scrolling


---

Note: there's the [`feed` ARIA role](https://www.w3.org/TR/wai-aria-1.1/#feed) for infinite scrolling, but it's not really supported by assistive technologies today

**Comment by samitron7** [comment]
- Arrow keys for navigation
**This sounds good to me**

- Don't mess with tabs 
**Also sounds good mainly because we should try to reserve tab from tabbing to next section and not within items of a section**

- load more" button instead of infinite scrolling
**//This is the only one I would like to find a better solution to but I say for now we can do the load more button. I've been tinkering with the idea of opening up a accessibility helper menu/mode where we can expose shortcut keys to help users move between large section without hitting the tab keys. Ie you enter menu and from menu you hit X key to go to Design panel or media panel inside of tabbing X times to nav between sections/actions. But I don't have time to explore this area now. @pbakaus Any thoughts on this?**

**Comment by pbakaus** [comment]
re: load more: I'm not sure this is the right compromise here.. but what we could try is that we don't trigger infinite scroll when the user is using a keyboard (instead, we only trigger infinite scroll if we detect that a) the trigger element at the bottom is visible and b) there were scroll events in a short time period before on the list...or something).

This is a little tricky because I'm not sure we can reliably distinguish scroll from keyboard and scroll from pan/mouse, but we could try.

**Comment by diegovar** [comment]
> > AFAIK you can't also suddenly ask them to use the Up/down/left/right keys when they happen to land in the list because they have no way of knowing that their method of input is supposed to change. So tabs should work.
> 
> I disagree. You **can** inform users that they can interact with a list using arrow keys, and you _should_ do that. In fact, it's pretty common and encouraged. See ["Managing Focus"](https://www.w3.org/TR/wai-aria-1.1/#managingfocus) section of the ARIA spec:
> 
> > When the container or its active descendant has focus, the user may navigate through the container by pressing additional keys, such as the arrow keys, to change the currently active descendant. Any additional press of the main navigation key (generally the TAB key) will move out of the container to the next widget.
> 
> You cannot expect user to tab through 30 items before they can get to the next important area on the screen. The [grid widget](https://www.w3.org/TR/wai-aria-practices-1.1/#grid) might be a suitable way for us to announce that navigation style, but there are others too. [The examples](https://www.w3.org/TR/wai-aria-practices-1.1/examples/grid/LayoutGrids.html) list "a set of search results" as a possible use case.
> 
> You are right about the infinite scrolling part. It's difficult to get this right in an accessible way. WordPress core is even considering [removing infinite scrolling](https://core.trac.wordpress.org/ticket/50105) for its media library. Some of the issues mentioned there:
> 
> * it's impossible or very hard for keyboard users to reach content placed after an infinite scrolling region
> * no audible feedback or instructions about how infinite scrolling works, the current and total number of items, or when new items get loaded
> 
> So my tl;dr is:
> 
> * Arrow keys for navigation
> * Don't mess with tabs
> * "load more" button instead of infinite scrolling
> 
> Note: there's the [`feed` ARIA role](https://www.w3.org/TR/wai-aria-1.1/#feed) for infinite scrolling, but it's not really supported by assistive technologies today

Thanks for those links and examples, I wasn't aware of the "composite" widgets in the Aria spec, looks like the right approach!
